civis/io/_tables.py

```diff
@@ -1310,168 +1292,148 @@ def _run_cleaning(file_ids, client, need_table_columns, headers, delimiter,
-    first_completed = done.pop()
-    try:
-        output_file = client.jobs.list_runs_outputs(
-            first_completed.job_id,
-            first_completed.run_id
-        )[0]
-    except IndexError:
-        raise CivisImportError(
-            "Unable to retrieve output from CSV preprocess "
-            f"job {first_completed.job_id} run {first_completed.run_id}"
-        )
-    detected_info = client.files.get(output_file.object_id).detected_info
-    table_columns = (detected_info['tableColumns'] if need_table_columns
-                     else None)
-    if headers is None:
-        headers = detected_info['includeHeader']
-    if delimiter is None:
-        delimiter = detected_info['columnDelimiter']
-
-    compression = detected_info['compression']
-
-    _check_all_detected_info(detected_info, headers, delimiter, compression,
-                             output_file.object_id)
-
-    cleaned_file_ids.append(output_file.object_id)
-
-    # Ensure that all results from files are correctly accounted for -
-    # Since concurrent.futures.wait returns two sets, it is possible
-    # That done contains more than one Future. Thus it is necessary to account
-    # for these possible completed cleaning runs while waiting on those which
-    # are still running.
-    for result in concurrent.futures.as_completed(done | still_going):
+    for fut in futures:
         try:
-            output_file = client.jobs.list_runs_outputs(
-                result.job_id,
-                result.run_id
-            )[0]
+            obj = client.jobs.list_runs_outputs(fut.job_id, fut.run_id)[0]
         except IndexError:
             raise CivisImportError(
                 "Unable to retrieve output from CSV preprocess "
-                f"job {result.job_id} run {result.run_id}"
+                f"job {fut.job_id} run {fut.run_id}"
             )
-        detected_info = client.files.get(output_file.object_id).detected_info
-        if need_table_columns:
-            file_columns = detected_info['tableColumns']
-            _check_column_types(table_columns, file_columns,
-                                output_file.object_id)
-
-        _check_all_detected_info(detected_info, headers, delimiter,
-                                 compression, output_file.object_id)
-        cleaned_file_ids.append(output_file.object_id)
+        f = client.files.get(obj.object_id)
+        files.append(
+            _File(id=f.id, name=f.name, detected_info=f.detected_info)
+        )
 
     if need_table_columns:
-        table_columns = _replace_null_column_names(table_columns)
+        table_columns, allow_inconsistent_headers = _check_column_types(files)
+    else:
+        table_columns = None
+        allow_inconsistent_headers = False
+
+    try:
+        headers = _check_detected_info(files, "includeHeader", headers)
+    except CivisImportError:
+        if allow_inconsistent_headers:
```

Comment:
How do you feel about this logic?  I see what you are trying to address but not sure that it covers all cases.  If you feel good about it, we can keep it!  Otherwise another option is to only do the headers and delimiter checks when the provided value is `None`.  Also happy to discuss offline.

Reply:
My potential blind spot here is that I don't have much background about the internal CSV preprocessing code, whose behavior I attempted to understand through my own testing via this civis-python layer (hence my long comment in `_check_column_types`). If you have potential problematic cases in mind, I'd be curious about what they might be and I'd like to address them. Otherwise I'd feel comfortable to merge this in for now and come back with future PRs if the bugs show up. :-)